### PR TITLE
e2e tests for complete/reject etc. C2D message APIs

### PR DESCRIPTION
**Pull request by bikamani**
[NOTE] Do not merge, blocked on service issue

e2e tests are added for different scenarios covering

ReceiveAsync with timeout, cancellationToken, 
AbandonAsync, 
RejectAsync, 
CompleteAsync
C2D send multiple messages
& few corner cases (Receiving Abandoned message)

### CODE REVIEW

File: e2e/test/MessageSendE2ETests.cs [hunk 1]
```diff
@@ -30,7 +30,8 @@ public MessageSendE2ETests()
         [TestMethod]
         public async Task Message_DeviceSendSingleMessage_Amqp()
         {
-            await SendSingleMessage(TestDeviceType.Sasl, Client.TransportType.Amqp_Tcp_Only).ConfigureAwait(false);
+            //await SendSingleMessage(TestDeviceType.Sasl, Client.TransportType.Amqp_Tcp_Only).ConfigureAwait(false);
+            await SendMultipleMessages(TestDeviceType.Sasl, Client.TransportType.Amqp_Tcp_Only).ConfigureAwait(false);
```
Comment (abhipsaMisra) [inline e2e/test/MessageSendE2ETests.cs]: Why is SendSingleMessage commented? This looks to be a _"testing"_ code that needs to be reverted.

File: e2e/test/MessageSendE2ETests.cs [hunk 2]
```diff
@@ -157,6 +158,21 @@ public async Task X509_DeviceSendSingleMessage_Http()
...
         }
 
+        public async static Task SendMultipleMessages(TestDeviceType type, Client.TransportType transport)
+        {
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(DevicePrefix, type).ConfigureAwait(false);
+
+            using (DeviceClient deviceClient = testDevice.CreateDeviceClient(transport))
+            {
+                await deviceClient.OpenAsync().ConfigureAwait(false);
+                for(int i = 0; i < 50; i++)
```
Comment (abhipsaMisra) [inline e2e/test/MessageSendE2ETests.cs]: The value 50 should be moved to a constant and a comment should be added to clarify why it is 50.

File: e2e/test/CombinedSendReceiveE2ETests.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            var task1 = Task.Run(async () =>
+            {
+                await MessageReceiveE2ETests.SendReceiveMultipleMessages(type, transport).ConfigureAwait(false);
+            });
+
+            var task2 = Task.Run(async () =>
+            {
+                await MessageSendE2ETests.SendMultipleMessages(type, transport).ConfigureAwait(false);
+            });
```
Comment (abhipsaMisra) [inline e2e/test/CombinedSendReceiveE2ETests.cs]: Why are task1 and task2 not awaited?

File: e2e/test/MessageSendE2ETests.cs (hunk 1 above)
Comment (bikamani) [reply e2e/test/MessageSendE2ETests.cs]: reverted, checking in now

File: e2e/test/MessageSendE2ETests.cs (hunk 2 above)
Comment (bikamani) [reply e2e/test/MessageSendE2ETests.cs]: Sure will add comment

File: e2e/test/MessageReceiveE2ETests.cs [hunk 4]
```diff
@@ -28,63 +32,147 @@ public MessageReceiveE2ETests()
...
         [TestMethod]
-        public async Task Message_DeviceReceiveSingleMessage_Http()
+        public async Task Message_DeviceReceiveSingleMessageReject_Http()
         {
-            await ReceiveSingleMessage(TestDeviceType.Sasl, Client.TransportType.Http1).ConfigureAwait(false);
+            await ReceiveSingleMessage(TestDeviceType.Sasl, Client.TransportType.Http1, VerifyReceivedC2DMessageAndReject).ConfigureAwait(false);
         }
 
         [TestMethod]
         public async Task X509_DeviceReceiveSingleMessage_Amqp()
```
Comment (abhipsaMisra) [inline e2e/test/MessageReceiveE2ETests.cs]: nit: Naming convention - these tests should also be renamed to indicate if the receive operation completes/is rejected/ is abandoned.

Comment (bikamani) [reply e2e/test/MessageReceiveE2ETests.cs]: Test name already reflects Reject/abandon etc. ReceiveSingleMessage is still relevant here as the test is receiving one message and then reject/abandon/complete it. Let me know if I misunderstood your comment.

File: e2e/test/CombinedSendReceiveE2ETests.cs (hunk 3 above)
Comment (bikamani) [reply e2e/test/CombinedSendReceiveE2ETests.cs]: Talked offline, will add appropriate changes

File: e2e/test/MessageReceiveE2ETests.cs [hunk 5]
```diff
@@ -28,63 +32,147 @@ public MessageReceiveE2ETests()
...
 
         [TestMethod]
         public async Task X509_DeviceReceiveSingleMessage_Http()
         {
-            await ReceiveSingleMessage(TestDeviceType.X509, Client.TransportType.Http1).ConfigureAwait(false);
+            await ReceiveSingleMessage(TestDeviceType.X509, Client.TransportType.Http1, VerifyReceivedC2DMessageAndComplete).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task Message_DeviceReceiveMultipleMessages_Amqp()
```
Comment (abhipsaMisra) [inline e2e/test/MessageReceiveE2ETests.cs]: nit: naming - same as above

Comment (bikamani) [reply e2e/test/MessageReceiveE2ETests.cs]: This was left due to some testing, test name is right, added right call.

File: e2e/test/MessageSendE2ETests.cs (hunk 1 above)
Comment (abhipsaMisra) [reply e2e/test/MessageSendE2ETests.cs]: What is the scenario that we are trying the verify here? If _send_ works for a single message, it would work for >1 messages as well.

Comment (bikamani) [reply e2e/test/MessageSendE2ETests.cs]: Adding @prmathur-microsoft , she has more context on why we need multiple message test.

File: e2e/test/MessageReceiveE2ETests.cs [hunk 6]
```diff
@@ -165,16 +335,126 @@ private async Task ReceiveSingleMessage(TestDeviceType type, Client.TransportTyp
...
+                CancellationTokenSource cts2 = new CancellationTokenSource(TimeSpan.FromSeconds(20));
+
+                /* DeviceQueue on service holds at max 50 messages which are not acked i.e. complete/reject/abandon 
+                 * from device Client, for receiving the 51st message without acking previous exception should be raised */
+                try
+                {
+                    for (int i = 0; i < 51; i++)
+                    {
+                        (Message msg1, string messageId1, string payload1, string p1Value) = ComposeC2DTestMessage();
+                        msg1.ExpiryTimeUtc = new DateTime(2019, 10, 24);
```
Comment (abhipsaMisra) [inline e2e/test/MessageReceiveE2ETests.cs]: Hardcoding the expiry time to 10/24/2019 will cause messages to be marked as expired (in the future). Use something like (CurrentTime + 1 hour) instead.

Comment (bikamani) [reply e2e/test/MessageReceiveE2ETests.cs]: agreed, good catch. will do

File: e2e/test/MessageReceiveE2ETests.cs [hunk 7]
```diff
@@ -165,16 +335,126 @@ private async Task ReceiveSingleMessage(TestDeviceType type, Client.TransportTyp
...
+                        await serviceClient.SendAsync(testDevice.Id, msg1).ConfigureAwait(false);
+                        numMessages++;
+                    }
+                }
+                catch (Exception ex)
```
Comment (abhipsaMisra) [inline e2e/test/MessageReceiveE2ETests.cs]: Add an additional Assertion to make sure that an Exception is thrown.

File: e2e/test/MessageSendE2ETests.cs (hunk 1 above)
Comment (bikamani) [reply e2e/test/MessageSendE2ETests.cs]: On a second thought, we test it in new e2e combined tests for parallel tests on send and receive. Hence removing these additional tests.

File: e2e/test/MessageReceiveE2ETests.cs [hunk 8]
```diff
@@ -165,16 +335,127 @@ private async Task ReceiveSingleMessage(TestDeviceType type, Client.TransportTyp
...
 
+                if (p1Value == prop2.Value)
+                {
+                    VerifyMessage(messageId1, payload1, p1Value, receivedMessage2);
+                    await deviceClient.CompleteAsync(receivedMessage2).ConfigureAwait(false);
+                }
+                else if (p1Value == prop3.Value)
+                {
+                    VerifyMessage(messageId1, payload1, p1Value, receivedMessage3);
+                    await deviceClient.CompleteAsync(receivedMessage3).ConfigureAwait(false);
```
Comment (abhipsaMisra) [inline e2e/test/MessageReceiveE2ETests.cs]: Shouldn't we be completing both the messages?

Comment (bikamani) [reply e2e/test/MessageReceiveE2ETests.cs]: yes moved it after if else

**Comment by bikamani** [comment]
Closing the PR for now until we are unblocked to get the tests passing.
